eureka/src/main/java/com/linecorp/armeria/client/eureka/EurekaEndpointGroup.java

```diff
@@ -388,6 +413,21 @@ private static Endpoint endpoint(InstanceInfo instanceInfo, boolean secureVip) {
         if (ipAddr != null && hostname != ipAddr) {
             endpoint = endpoint.withIpAddr(ipAddr);
         }
+        if (instanceMetadataAsAttrs) {
+            endpoint = withInstanceMetadata(endpoint, instanceInfo);
+        }
+
         return endpoint;
     }
+
+    private static Endpoint withInstanceMetadata(Endpoint endpoint, InstanceInfo instanceInfo) {
+        final List<Map.Entry<AttributeKey<?>, ?>> attrs = new ArrayList<>(instanceInfo.getMetadata().size());
+        attrs.add(new SimpleImmutableEntry<>(INSTANCE_INFO, instanceInfo));
+        instanceInfo.getMetadata().forEach((key, value) -> {
+            if (!"@class".equals(key)) {
+                attrs.add(new SimpleImmutableEntry<>(AttributeKey.valueOf(key), value));
```

Comment:
This adds just String value from the metadata map.
Instead of adding each entry, how about just adding the whole metadata at once using a separate key?

Reply:
Or Could we just add `instanceInfo` and let users extract the metadata from the `instanceInfo`?
In that case we need to make the class public.

Reply:
I prefer adding a common  API for all discovery Endpoints. Not just add `instanceInfo`.

Reply:
> I prefer adding a common API for all discovery Endpoints.

Yes, exactly. 😄 This is for `EurekaEndpointGroup` not for other `EndpointGroup`.
Users who use `EurekaEndpointGroup` would be able to extract `instanceInfo` and do the extra work with it.
Also instead of adding too many `Attributes` by spreading the entries of `metadata`, we can just use one key so that users can easily check the content of the metadata.

Reply:
If we are really reluctant to make this public, how about using just one key for the metadata?
```java
private static final AttributeKey<Map<?, ?>> METADATA = AttributeKey.valueOf(
            EurekaEndpointGroup.class, "METADATA");
```

Reply:
metadata is common in many registry clients. But the client API may differ from each other.  
Users may want to use a common API to get the value.

Spring cloud add `Map<String, String> getMetadata();` to serviceinstance.
https://github.com/spring-cloud/spring-cloud-commons/blob/8807b4dac035c2247987d22065ac03b8c86777e3/spring-cloud-commons/src/main/java/org/springframework/cloud/client/ServiceInstance.java#L65

That is what we want to do in this MR at the beginning.  If we add the whole metadata using a separate key.
We may need to add another API. 
```
public final class InstanceInfoUtil {
    
    Map<String, String> getMetadata() getMetadata(Endpoint endpoint) {
        return endpoint.attr(..);
    }
}
```
I

Reply:
> We may need to add another API.

Yes, that's actually what I expected.
```java
public final class MetadataUtil {
    private static final AttributeKey<Map<?, ?>> METADATA = AttributeKey.valueOf(...);

    Map<String, String> getMetadata() getMetadata(Endpoint endpoint) {
        return endpoint.attr(METADATA);
    }
}
```
I think it's pretty common so that we can add it and may use it for other discovery `EndpointGroup` as well.
But at this time, we are not sure if it's really needed for other discoveries so we can just put the `MetadataUtil` class into the `eureka` module and expand it later when we need.

But before that, we have to make sure if it's a good idea to add `InstanceInfo` to the `Endpoint`:
https://github.com/line/armeria/pull/3980/files#diff-7bf2b0d8d123c011e14425573a7aad2a7e9a9847fd188d63cbba79d58f7969a9R425
If it's just an internal API, we don't need to add it because it isn't used anywhere.

So now, we have to decide on these things:
- Making `instanceInfo` public or not adding it to the `Endpoint`
  - If we make `InstanceInfo` public, then we don't need to add `metadata` separately.
  - If we don't add `InstanceInfo` to the `Endpoint`, we have to add `metadata` to the `Endpoint` as attributes
    - We can add `metadata` as once using a specific key for it
    - Or spread `metadata` and add `entry`s one by one.

Reply:
But in order to keep the immutability of the endpoint, does it means that we have to keep the `Map<String, String> ` return by metadata
immutable? That is why we replace AttributeMap with `withAttr` and `attr`.
I think we should add entries one by one.

Reply:
> does it means that we have to keep the Map<String, String> return by metadata
immutable?

I think so and we should. If we just set the immutable metadata to the `Endpoint`, there's no problem. I thought we were setting the immutable metadata when we build an `InstanceInfo` but apparently, we were not doing that. 😅 We should make a change.

But I'm still leaning toward making the `InstanceInfo` public. Users who use `EurekaEndpointGroup` can leverage `InstanceInfo` for fine-grained control CSLB when it's necessary.

Reply:
I slightly prefer just exposing `InstanceInfo` as a public API, but it may make this PR a little bit bigger than initially intended. How about splitting this PR into two, so we can focus on adding attributes to `Endpoint` first, and then making `InstanceInfo` (and its related classes) public?